### PR TITLE
FIX #162 "attached to a different loop", and add a new test case

### LINKED ISSUE #162: Teardown error: "attached to a different loop" with even_loop module scope fixture
I'm working on python/pytest/asyncio since beginning of this year, and when updating pytest-asyncio (from 0.10.0 to 0.12.0) then some previous working tests started to fail.

I have created following minimal example (isolated from a more complex solution) that always fails at teardown.
Can you please let me know if you can reproduce it, and your kind feedback about this potential issue.  
Don´t hesitate to ask for any additional information or test on my side.

BR

Angel Luis

# Additional info
## Versions used to reproduce are: 

* pytest-asyncio 0.12.0. (it fails also with 0.11.0)
* pytest 5.4.2
* python 3.8.2
* setuptools 46.2.0 & pip 20.1
* windows 10 Pro [Versión 10.0.18362.175]
Note: Minimal Example works with pytest-asyncio 0.10.0. 

## Minimal example is the following
```
import asyncio
import contextlib
import functools
import pytest


@pytest.mark.asyncio
async def test_simple(port):
    assert True

@pytest.fixture(scope="module")
def event_loop():
    """Change event_loop fixture to module level."""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()


@pytest.fixture(scope="module")
async def port(request, event_loop):
    def port_finalizer(finalizer):
        async def port_afinalizer():
            await finalizer(None, None, None)
        event_loop.run_until_complete(port_afinalizer())

    context_manager = port_map()
    port = await context_manager.__aenter__()
    request.addfinalizer(functools.partial(port_finalizer, context_manager.__aexit__))
    return port

@contextlib.asynccontextmanager
async def port_map():
    worker = asyncio.create_task(background_worker())
    yield
    try:
        worker.cancel()
        await worker
    except asyncio.CancelledError:
        pass

async def background_worker():
    while True:
        await asyncio.sleep(10.0)

```
When I run the test I get the following console output: 

```
(venv) C:\git\ATB2\ejemplo_issue>pytest
=================== test session starts =============================
platform win32 -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: C:\git\ATB2\ejemplo_issue
plugins: asyncio-0.12.0
collected 1 item                                                                                                                                                                                                                                                           

test_example.py .E                                                                                                                                                                                                                                                   [100%]

====================== ERRORS =======================================
______________ ERROR at teardown of test_simple _____________________

finalizer = <bound method _AsyncGeneratorContextManager.__aexit__ of <contextlib._AsyncGeneratorContextManager object at 0x0423AC40>>

    def port_finalizer(finalizer):
        async def port_afinalizer():
            await finalizer(None, None, None)
>       event_loop.run_until_complete(port_afinalizer())

test_example.py:24:
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\ablasco\AppData\Local\Programs\Python\Python38-32\lib\asyncio\base_events.py:616: in run_until_complete
    return future.result()
test_example.py:23: in port_afinalizer
    await finalizer(None, None, None)
C:\Users\ablasco\AppData\Local\Programs\Python\Python38-32\lib\contextlib.py:178: in __aexit__
    await self.gen.__anext__()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @contextlib.asynccontextmanager
    async def port_map():
        worker = asyncio.create_task(background_worker())
        yield
        try:
            worker.cancel()
>           await worker
E           RuntimeError: Task <Task pending name='Task-4' coro=<port.<locals>.port_finalizer.<locals>.port_afinalizer() running at C:\git\ATB2\ejemplo_issue\test_example.py:23> cb=[_run_until_complete_cb() at C:\Users\ablasco\AppData\Local\Programs\Python\Python38-32
\lib\asyncio\base_events.py:184]> got Future <Task pending name='Task-2' coro=<background_worker() running at C:\git\ATB2\ejemplo_issue\test_example.py:43> wait_for=<Future cancelled>> attached to a different loop

test_example.py:37: RuntimeError
============================= short test summary info ==============================
ERROR test_example.py::test_simple - RuntimeError: Task <Task pending name='Task-4' coro=<port.<locals>.port_finalizer.<locals>.port_afinalizer() running at C:\git\ATB2\ejemplo_issue\test_example.py:23> cb=[_run_until_complete_cb() at C:\Users\ablasco\AppData\Local...

=========================== 1 passed, 1 error in 0.13s =============================

```
## Unexpectedly either of the following changes, made the test to pass:

a) Change fixture to a lower level scope (class or function): 
```
@pytest.fixture(scope="class")
def event_loop():
```
```
@pytest.fixture(scope="class")
async def port(request, event_loop):
```
b) Embed the test inside a class:
```
class TestClass:
    @pytest.mark.asyncio
    async def test_simple(port):
        assert True
```
c) Or change finalizer to call get_event_loop (instead of using event_loop from fixture)
```
    def port_finalizer(finalizer):
        async def port_afinalizer():
            await finalizer(None, None, None)
        # event_loop.run_until_complete(port_afinalizer())
        loop = asyncio.get_event_loop()
        loop.run_until_complete(port_afinalizer())
```

### PR DESCRIPTION
**Pull request by alblasco**
1) A new test case that fails with 0.12.0, and pass with this commit.
   test_async_fixtures_with_finalizer_scope.py:
2) This commit fixes issue https://github.com/pytest-dev/pytest-asyncio/issues/162
   Main problem is due to side effects of asyncio.get_event_loop(). See:
   https://github.com/python/cpython/blob/3.8/Lib/asyncio/events.py#L636
   This method could either return an existing loop or create a new one.

This commit replaces all asyncio.get_event_loop() in order to use the loop provided by event_loop fixture (instead of calling asyncio.get_event_loop())
*  with previous pytest-asyncio code 0.10.0 (mostly undoing changes from https://github.com/pytest-dev/pytest-asyncio/commit/a9e2213224151f35a695dd588d3bb79c42ff9f3b, that makes a wrong assumption), 
* apply similar to wrap_in_sync
* Except following block, that has not been modified in this commit (as it behaves similar to 0.10.0)
   https://github.com/pytest-dev/pytest-asyncio/blob/v0.12.0/pytest_asyncio/plugin.py#L54-L66

Changes are for using always the new loop provided by event_loop fixture.
   Instead of calling get_event_loop() that:
   - either returns global recorded loop (with set_event_loop()) in case there is one,
   - or otherwise creates and record a new one

BTW https://github.com/pytest-dev/pytest-asyncio/commit/a9e2213224151f35a695dd588d3bb79c42ff9f3b is not on the 0.10.0 changelog

### CODE REVIEW

File: tests/async_fixtures/test_async_fixtures_with_finalizer_scope.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            await finalizer(None, None, None)
+        event_loop.run_until_complete(port_afinalizer())
+
+    context_manager = port_map()
+    port = await context_manager.__aenter__()
+    request.addfinalizer(functools.partial(port_finalizer, context_manager.__aexit__))
+    return True
+
+
+@contextlib.asynccontextmanager
```
Comment (yanlend) [inline tests/async_fixtures/test_async_fixtures_with_finalizer_scope.py:34]: This seems to fail the build on older Python versions. I don't think it is an essential part of the test. So you might get it to run on all required versions by changing the test a bit.

File: pytest_asyncio/plugin.py [hunk 2]
```diff
@@ -96,21 +103,30 @@ async def async_finalizer():
...
 
             request.addfinalizer(finalizer)
-            return asyncio.get_event_loop().run_until_complete(setup())
+            return loop.run_until_complete(setup())
 
         fixturedef.func = wrapper
     elif inspect.iscoroutinefunction(fixturedef.func):
         coro = fixturedef.func
 
+        strip_event_loop = False
```
Comment (yanlend) [inline pytest_asyncio/plugin.py:115]: This code is copy pasted from line 72. Can you refactor it such that it is needed only once? E.g. pull it into a small function.

**Comment by yanlend** [comment]
Great fix, thank you! Please review and merge this.

**Comment by Tinche** [comment]
Hello, thanks for this! The next few days are jam packed for me so this might wait until the end of the week for me to review. Hopefully the community can also comment in the mean time.

File: tests/async_fixtures/test_async_fixtures_with_finalizer_scope.py (hunk 1 above)
Comment (alblasco) [reply tests/async_fixtures/test_async_fixtures_with_finalizer_scope.py]: Fixed.

File: pytest_asyncio/plugin.py (hunk 2 above)
Comment (alblasco) [reply pytest_asyncio/plugin.py]: Fixed. Included a new class FixtureStripper. 

Please review again as I have included a new test case, and this has required additional changes
Thanks in advance.

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    yield loop
+    loop.close()
+
+
+@pytest.fixture(scope="module")
+async def port1(request, event_loop):
+    def port_finalizer(finalizer):
+        async def port_afinalizer():
+            # await task inside get_event_loop()
+            # RantimeError is raised if task is created on a different loop
```
Comment (yanlend) [inline tests/async_fixtures/test_async_fixtures_with_finalizer.py:31]: Typo: RuntimeError

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    loop = policy.new_event_loop()
+    yield loop
+    loop.close()
+
+
+@pytest.fixture(scope="module")
+async def port1(request, event_loop):
+    def port_finalizer(finalizer):
+        async def port_afinalizer():
+            # await task inside get_event_loop()
```
Comment (yanlend) [inline tests/async_fixtures/test_async_fixtures_with_finalizer.py:30]: This task is not awaited inside `get_event_loop()`, but in `loop.run_until_complete()`.

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    request.addfinalizer(functools.partial(port_finalizer, worker))
+    return True
+
+
+@pytest.fixture(scope="module")
+async def port2(request, event_loop):
+    def port_finalizer(finalizer):
+        async def port_afinalizer():
+            # await task inside get_event_loop()
+            # if loop is different a RuntimeError is raised
```
Comment (yanlend) [inline tests/async_fixtures/test_async_fixtures_with_finalizer.py:45]: I would suggest to align the two comments from `port1` and `port2` and then highlight the difference of the two fixtures (first uses `event_loop`, second uses `asyncio.get_event_loop`)

File: pytest_asyncio/plugin.py (hunk 2 above)
Comment (yanlend) [reply pytest_asyncio/plugin.py]: Had some minor comments, I am happy overall.

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py (hunk 3 above)
Comment (alblasco) [reply tests/async_fixtures/test_async_fixtures_with_finalizer.py]: Fixed

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py (hunk 4 above)
Comment (alblasco) [reply tests/async_fixtures/test_async_fixtures_with_finalizer.py]: Fixed

File: tests/async_fixtures/test_async_fixtures_with_finalizer.py (hunk 5 above)
Comment (alblasco) [reply tests/async_fixtures/test_async_fixtures_with_finalizer.py]: Done

**Comment by Tinche** [comment]
OK let's get this in. Could you add a line to the changelog?

**Comment by alblasco** [comment]
> OK let's get this in. Could you add a line to the changelog?

Line added. Let me know if anything else is needed

**Comment by Tinche** [comment]
Thanks a lot, release going out tomorrow!

**Comment by jack1142** [comment]
> Thanks a lot, release going out tomorrow!

@Tinche a friendly ping reminder as you said "tomorrow" some time ago and there's been no release :)
Please ignore me if you're busy, I just wanted to remind you about it in case you forgot.

**Pull request merged**